File: src-tauri/src/status.rs
use serde::Serialize;
use sqlx::SqlitePool;
use tauri::TitleBarStyle;

const LAST_SYNC_KEY: &str = "last_sync_ms";

#[derive(Debug, Clone, Serialize)]
pub struct AppStatus {
    /// Email addresses of connected accounts; empty means "not signed in".
    pub accounts: Vec<String>,
    /// The subset of `accounts` whose refresh token is dead for good
    /// (`AppState::reauth`): sync has stopped trying them, and the only fix
    /// is signing in again. The UI turns a non-empty list into a reconnect
    /// prompt — which is why this is emails rather than a count: "an account
    /// needs attention" with two accounts connected is a guessing game.
    pub needs_reauth: Vec<String>,
    pub last_sync_ms: Option<i64>,
    /// True when the app is running on synthetic data, so the UI can say so.
    pub demo: bool,
    /// True when the window's own controls are drawn *over* the webview rather
    /// than in a strip above it, so the header has to leave room for them.
    ///
    /// This rides on `get_status` for one reason: `ui/src` has no other way to
    /// learn what it is running on. Nothing in the frontend is platform-aware,
    /// and the alternative — `@tauri-apps/plugin-os` — is a whole dependency
    /// bought for a single boolean the app already fetches a payload for on
    /// mount. It sits beside `demo` rather than in `get_palette` because it is
    /// the same kind of fact: not a colour, but something about *this* running
    /// instance that the UI has to render differently for.
    pub overlay_titlebar: bool,
}

/// Whether the window controls overlay the webview's own content.
///
/// Both arguments, rather than reading `cfg!` here: the config half is what the
/// shipped `tauri.conf.json` actually asks for (so removing `titleBarStyle`
/// removes the inset with it, instead of leaving the UI reserving space for
/// controls that moved back into their own strip), and the platform half is what
/// keeps Omarchy out of it. `titleBarStyle` is a macOS-only key — Linux parses
/// it and ignores it — so on Linux the config says `Overlay` and the controls
/// are still in a strip of their own. Reserving 60px there buys a dead gap at
/// the left of the header and nothing else.
pub fn controls_overlay_content(style: TitleBarStyle, macos: bool) -> bool {
    macos && matches!(style, TitleBarStyle::Overlay)
}

pub async fn read_status(
    pool: &SqlitePool,
    demo: bool,
    overlay_titlebar: bool,
    needs_reauth: Vec<String>,
) -> anyhow::Result<AppStatus> {
    let accounts: Vec<String> =
        sqlx::query_scalar("SELECT email FROM accounts ORDER BY id")
            .fetch_all(pool)
            .await?;

    Ok(AppStatus {
        accounts,
        needs_reauth,
        last_sync_ms: last_sync_ms(pool).await?,
        demo,
        overlay_titlebar,
    })
}

/// When the last successful sync was recorded, and nothing else.
///
/// The background loop wants this one number to decide whether a sync is due.
/// It has no use for the account list, and — more to the point — no honest
/// value for either of `read_status`'s flags: it is not the UI, so whether the
/// window controls overlay the content is none of its business, and passing a
/// placeholder for a field somebody later reads is how a placeholder becomes a
/// claim.
pub async fn last_sync_ms(pool: &SqlitePool) -> anyhow::Result<Option<i64>> {
    let raw: Option<String> =
        sqlx::query_scalar("SELECT value FROM settings WHERE key = ?1")
            .bind(LAST_SYNC_KEY)
            .fetch_optional(pool)
            .await?;
    Ok(raw.and_then(|v| v.parse().ok()))
}

pub async fn record_sync(pool: &SqlitePool, at_ms: i64) -> anyhow::Result<()> {
    sqlx::query(
        "INSERT INTO settings (key, value) VALUES (?1, ?2)
         ON CONFLICT (key) DO UPDATE SET value = excluded.value",
    )
    .bind(LAST_SYNC_KEY)
    .bind(at_ms.to_string())
    .execute(pool)
    .await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    async fn pool_with_account() -> sqlx::SqlitePool {
        let pool = omacal_store::connect_memory().await.unwrap();
        sqlx::query(
            "INSERT INTO accounts (google_sub, email, created_at) VALUES ('s','me@x.com',0)")
            .execute(&pool).await.unwrap();
        pool
    }

    #[tokio::test]
    async fn status_reports_no_accounts_on_a_fresh_database() {
        let pool = omacal_store::connect_memory().await.unwrap();
        let s = read_status(&pool, false, false, vec![]).await.unwrap();
        assert!(s.accounts.is_empty());
        assert_eq!(s.last_sync_ms, None);
        assert!(!s.demo);
    }

    #[tokio::test]
    async fn status_lists_connected_accounts_by_email() {
        let pool = pool_with_account().await;
        let s = read_status(&pool, false, false, vec![]).await.unwrap();
        assert_eq!(s.accounts, vec!["me@x.com".to_string()]);
    }

    #[tokio::test]
    async fn recording_a_sync_round_trips() {
        let pool = pool_with_account().await;
        record_sync(&pool, 1_785_715_200_000).await.unwrap();
        let s = read_status(&pool, false, false, vec![]).await.unwrap();
        assert_eq!(s.last_sync_ms, Some(1_785_715_200_000));
    }

    #[tokio::test]
    async fn recording_a_sync_twice_keeps_the_latest() {
        let pool = pool_with_account().await;
        record_sync(&pool, 1_000).await.unwrap();
        record_sync(&pool, 2_000).await.unwrap();
        let s = read_status(&pool, false, false, vec![]).await.unwrap();
        assert_eq!(s.last_sync_ms, Some(2_000));
    }

    /// The reconnect list rides through untouched — and as emails, because
    /// "an account needs attention" with two accounts connected is a guessing
    /// game the UI cannot resolve on the user's behalf.
    #[tokio::test]
    async fn status_surfaces_the_accounts_needing_reconnection() {
        let pool = pool_with_account().await;

        let s = read_status(&pool, false, false, vec!["me@x.com".into()]).await.unwrap();
        assert_eq!(s.needs_reauth, vec!["me@x.com".to_string()]);

        let none = read_status(&pool, false, false, vec![]).await.unwrap();
        assert!(none.needs_reauth.is_empty(), "a healthy account was told to reconnect");
    }

    /// Both flags at once, and crossed: `demo` and `overlay_titlebar` are
    /// adjacent `bool` parameters, so the one mistake this signature invites is
    /// passing them the wrong way round. Asserting each in the run where the
    /// *other* is false is what makes that swap fail rather than cancel out.
    #[tokio::test]
    async fn the_demo_and_overlay_flags_are_surfaced_independently() {
        let pool = omacal_store::connect_memory().await.unwrap();

        let demo_only = read_status(&pool, true, false, vec![]).await.unwrap();
        assert!(demo_only.demo, "the demo badge would never appear");
        assert!(!demo_only.overlay_titlebar, "demo leaked into the titlebar flag");

        let overlay_only = read_status(&pool, false, true, vec![]).await.unwrap();
        assert!(overlay_only.overlay_titlebar, "the header would never clear the traffic lights");
        assert!(!overlay_only.demo, "the titlebar flag leaked into the demo badge");
    }

    /// The whole of the Linux half of this feature, and the half no macOS run
    /// can observe. `titleBarStyle` is macOS-only, so `tauri.conf.json` says
    /// `Overlay` on Omarchy too — and there the controls still sit in a strip of
    /// their own, so an inset is a dead ~60px gap at the left of the header.
    #[test]
    fn overlay_reserves_room_only_on_macos() {
        assert!(controls_overlay_content(TitleBarStyle::Overlay, true));
        assert!(
            !controls_overlay_content(TitleBarStyle::Overlay, false),
            "Linux would get a dead gap where macOS has traffic lights"
        );
    }

    /// The other direction: macOS alone is not the reason to inset. A window
    /// whose controls are back in their own strip needs no room made for them,
    /// and reserving it would put the title 60px adrift of nothing.
    #[test]
    fn a_normal_title_bar_reserves_no_room_on_either_platform() {
        for macos in [true, false] {
            assert!(!controls_overlay_content(TitleBarStyle::Visible, macos), "macos={macos}");
            assert!(!controls_overlay_content(TitleBarStyle::Transparent, macos), "macos={macos}");
        }
    }

    /// What makes the two bands one on macOS, and the one thing here no other
    /// test can see: `get_status` derives `overlay_titlebar` from this file, so
    /// a config that stopped asking for `Overlay` would take the inset with it
    /// and nothing would be wrong — except that the title strip would be back.
    ///
    /// `include_str!`, so the path is checked when this compiles rather than
    /// when it runs, and no test needs a working directory.
    #[test]
    fn the_shipped_window_config_asks_for_a_unified_title_bar() {
        let conf: serde_json::Value =
            serde_json::from_str(include_str!("../tauri.conf.json")).unwrap();
        let w = &conf["app"]["windows"][0];

        assert_eq!(w["titleBarStyle"], "Overlay", "the title bar is back in its own strip");
        assert_eq!(w["hiddenTitle"], true, "\"omacal\" is drawn over the header again");
        // Not a round number for its own sake: 800x600 cannot show a working
        // week and a month grid without crushing both.
        assert_eq!(w["width"], 1280, "the default window is back to a size no calendar fits");
        assert_eq!(w["height"], 800, "the default window is back to a size no calendar fits");
    }
}

File: ui/src/lib/status.ts
import { invoke } from '@tauri-apps/api/core';

export type AppStatus = {
  accounts: string[];
  /** The subset of `accounts` whose Google sign-in is dead for good — the
   *  token endpoint said so, sync has stopped trying them, and signing in
   *  again is the only fix. Emails rather than a count, because "an account
   *  needs attention" with two connected is a guessing game. */
  needs_reauth: string[];
  /** A newer published release, when the backend's daily check found one —
   *  the header shows a quiet notice for it. Null means current, unchecked,
   *  or demo; the UI has no reason to tell those apart. */
  update: { version: string; url: string } | null;
  /** The running build's own version, shown in Settings. Same source the
   *  update check compares against, so the two can never disagree. */
  version: string;
  /** The system time zone's new IANA name, when it moved out from under this
   *  process — whose own zone (the one every time on screen is drawn in) is
   *  fixed at launch. Null means nothing moved, or the display zone is
   *  pinned by setting and cannot go stale. The header turns this into the
   *  banner whose one action is the restart that catches the app up. */
  system_tz_change: string | null;
  last_sync_ms: number | null;
  demo: boolean;
  /** True when the window's controls are drawn over the webview instead of in
   *  a strip above it, so the header has to leave room for them — macOS with
   *  `titleBarStyle: "Overlay"`, and nothing else.
   *
   *  This is the only thing `ui/src` knows about the platform it runs on, and
   *  it deliberately arrives as a *property of the window* rather than as an
   *  OS name: "leave room at the left" is what the header needs to decide, and
   *  the OS is only one of the two things that decides it (see
   *  `status::controls_overlay_content` for the other). It rides on
   *  `get_status` because that is already fetched on mount; the alternative
   *  was `@tauri-apps/plugin-os`, a dependency for one boolean. */
  overlay_titlebar: boolean;
};

export const getStatus = () => invoke<AppStatus>('get_status');
export const signIn = () => invoke<string>('sign_in');
export const syncNow = () => invoke<number>('sync_now');
/** Opens the latest release's page in the system browser. No argument on
 *  purpose: the backend opens the URL *it* fetched, so the webview never
 *  chooses what the browser is pointed at. */
export const openLatestRelease = () => invoke<void>('open_latest_release');
/** Restarts the app — the moved-zone banner's one action. The backend delays
 *  the re-exec long enough for this call to resolve, so the button can say
 *  "Restarting…" instead of dying mid-await. */
export const restartApp = () => invoke<void>('restart_app');

/**
 * What the header's status light is showing.
 *
 * Four states, because four is what a person can tell apart at 8px without a
 * label — spec §2's table, named.
 */
export type SyncState =
  /** A sync is in flight. Active. */
  | 'syncing'
  /** Something failed. The one state that has to be noticed. */
  | 'failed'
  /** Nothing to be stale: signed out, or signed in and nothing fetched yet. */
  | 'never'
  /** The normal case, and the one that must be quiet enough to ignore. */
  | 'synced';

/**
 * The status light's whole answer: which state, and the sentence that says so.
 *
 * **One call for both, and that is the design rather than a convenience.**
 * Spec §2 requires that a colour is never the only signal, and the way to
 * guarantee it is not to remember to write the words — it is to make the
 * colour and the words the same answer. The component derives its class *and*
 * its `aria-label`/`title` from this one result, so a dot showing one thing
 * while announcing another is not a state this code can reach.
 *
 * The order the facts are read in is the only real decision here, and each
 * step is a case where two things are true at once:
 *
 *   1. **`busy` first.** A sync in flight outranks the failure it is retrying;
 *      otherwise the light stays red through the very moment it is being fixed.
 *   2. **`error` before `lastSyncMs`.** `lastSyncMs` is the last *successful*
 *      sync and survives a failed one, so reading it first reports "Synced
 *      5 min ago" while the calendar quietly goes stale — the exact defect a
 *      status light exists to make visible.
 *   3. **Signed out is not a failure.** Nothing has gone wrong; there is
 *      nothing to sync. A red dot on a fresh install is the app telling a new
 *      user it is broken.
 *
 * `error` is `App`'s own error state, which is wider than sync: a write that
 * could not refresh sets it too, and `AppStatus` carries no sync-specific
 * failure field to narrow to. Keeping the wider signal is deliberate — it is
 * the only failure the header is given, and under-reporting one is worse than
 * occasionally over-reporting.
 *
 * **So the label does not claim it was a sync failure.** Saying "Last sync
 * failed" when a *write* could not refresh states as fact something the source
 * cannot support: over-reporting that a problem exists is defensible,
 * misreporting which problem it is is not. The dot says something is wrong and
 * the sentence says what, by carrying `App`'s own message rather than a
 * category invented here.
 *
 * If it ever proves noisy — a transient write failure leaving the light red
 * until the next operation — that is when a sync-specific field on `AppStatus`
 * earns its cost. Not before.
 */
export function syncLight(
  o: {
    connected: boolean; busy: boolean; error: string | null; reauth: boolean;
    lastSyncMs: number | null;
  },
  now: number,
): { state: SyncState; label: string } {
  if (o.busy) return { state: 'syncing', label: 'Syncing now' };
  if (o.error !== null) return { state: 'failed', label: `Something went wrong: ${o.error}` };
  // After `error`: an account parked behind a reconnect is a standing state,
  // and a fresh failure is news. Before `lastSyncMs`, or the light reads
  // "Synced 5 min ago" in green while an account it stopped syncing goes
  // quietly stale — the exact defect the `error` ordering above exists to
  // avoid, one row down. This is the sync-specific status field the comment
  // above said would one day earn its cost.
  if (o.reauth) return { state: 'failed', label: 'An account needs to be reconnected' };
  if (!o.connected) return { state: 'never', label: 'Not signed in' };
  if (o.lastSyncMs === null) return { state: 'never', label: 'Not synced yet' };
  return { state: 'synced', label: `Synced ${relativeTime(o.lastSyncMs, now)}` };
}

/**
 * The reconnect banner's sentence. Names the account, because with two
 * connected the user's next action differs by which one is dead; "is no
 * longer valid" rather than "expired", because a revoked grant and a
 * client-pair change both land here and neither is about time.
 */
export function reauthMessage(emails: string[]): string {
  return `Google sign-in for ${emails.join(', ')} is no longer valid. Reconnect to resume syncing.`;
}

/**
 * The moved-zone banner's sentence. Both zones by name, because the damage
 * is the *difference*: "your time zone changed" on a machine whose clock
 * looks fine reads as a riddle, while naming the zone still on screen says
 * exactly what every hour on the grid currently means. `shownZone` is the
 * webview's own frozen zone — the header already knows it — so the sentence
 * cannot drift from what the grid is actually drawing.
 */
export function tzChangeMessage(newZone: string, shownZone: string): string {
  return `This machine moved to ${newZone}, but times are still shown in ${shownZone}. Restart OmaCal to catch up.`;
}

/** "just now" / "4 min ago" / "2 h ago" — deliberately coarse. */
export function relativeTime(ms: number | null, now = Date.now()): string {
  if (ms === null) return 'never';
  const s = Math.max(0, Math.floor((now - ms) / 1000));
  if (s < 60) return 'just now';
  if (s < 3600) return `${Math.floor(s / 60)} min ago`;
  if (s < 86400) return `${Math.floor(s / 3600)} h ago`;
  return `${Math.floor(s / 86400)} d ago`;
}
